### ISSUE TITLE
Copies to/from message payloads should use the sizeof() operator where possible

### ISSUE THREAD

**Issue by skliper** [nasa/cFE#144]
When reading/writing software bus message payloads, strings must be always char[] arrays (not pointers) by definition.  Many times these these are defined as something like:
 `char MyString[SIZE_MACRO];`

and later they are filled using something like:
 `strncpy(Message->MyString, Source, SIZE_MACRO);`

However, instead of using the `SIZE_MACRO` in the runtime code, it is more future-proof to use the `sizeof()` operator to get the array size.  This is particularly true when the payload structures are generated from an EDS (electronic data sheet) since there may not even be a simple macro to use in that case.  

The `sizeof()` operator will always work even if the size is a hardcoded value or the macro name changes in a future version.

**Comment by skliper** [comment]
_Imported from trac issue 113.  Created by jphickey on 2015-10-06T10:19:03, last modified: 2019-03-05T14:57:55_

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-10-06 10:58:47:_

Pushed branch `trac-113-use-sizeof-for-message-strings`, commit [changeset:2cdc4ab] for review

* Confirmed that all CFE unit tests pass with this change set (on 32-bit build, anyway).
* Also modifies one test case to create a software bus message where a string will intentionally //not// be terminated.  The ES code should handle this.  Unit tests were being too "nice" here by always correctly null terminating their messages.

**Comment by skliper** [comment]
_Trac comment by glimes on 2015-10-06 12:02:55:_

Having debugged problems in the past where the operand for a `sizeof` operator was suddenly morphed from an array to a pointer (usually via a typedef change done years later), I recommend that reviewers keep a careful eye on the objects being sized, to assure that they are not only arrays today but will continue to be arrays into the future.

Going to go continue my review, but since this is a short-fuse review I wanted to draw some attention to this potential for a future problem.

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-10-06 12:06:40:_

Replying to [comment:2 glimes]:
> Having debugged problems in the past where the operand for a `sizeof` operator was suddenly morphed from an array to a pointer (usually via a typedef change done years later), I recommend that reviewers keep a careful eye on the objects being sized, to assure that they are not only arrays today but will continue to be arrays into the future.


Agreed on that -- in this case these are software bus messages, so in fact having a pointer in here would break things horribly.  The `sizeof()` operations proposed here are limited to software bus messages for this reason.

**Comment by skliper** [comment]
_Trac comment by glimes on 2015-10-06 12:43:03:_

recommend accept.

side notes:

I see a couple of places where `strncpy(FOO, something, sizeof(FOO))` is used to replace code that previously included logic to assure FOO ends up containing a `\0` terminator. This could result in the destination buffer containing a string that is not `\0` terminated.

Have we finished fixing up any code that presumes the strings in these small fixed size arrays in the messages is necessarily `\0` terminated?

Thinking about the DRY principle ... `sizeof FOO ? sizeof FOO[0]` and `strncpy(FOO, ..., sizeof (FOO))` are both fairly common idioms, but when `FOO` is a fairly long expression they are a fertile source of potential bugs. Do we want to think about folding them up into macros, so `FOO` only appears once in the source code?

... ditto `snprintf(FOO, sizeof (FOO), ...)` and yes, we already use varargs macros so we may as well embrace them ;)

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-10-06 13:06:58:_

Replying to [comment:4 glimes]:
> I see a couple of places where `strncpy(FOO, something, sizeof(FOO))` is used to replace code that previously included logic to assure FOO ends up containing a `\0` terminator. This could result in the destination buffer containing a string that is not `\0` terminated.

This was most likely the unit tests.  IMO unit tests were being overly nice by ensuring all these null terminators.  A buggy ground station might not be so nice, also I don't know anywhere that we specify that strings within software bus messages are guaranteed to be null terminated, nor should any software reading data from outside assume anything about the format of that data anyway (bad idea).

So in short, it is intentional that the unit tests no longer guarantee that the messages will be null terminated.

**Comment by skliper** [comment]
_Trac comment by glimes on 2015-10-13 11:28:52:_

Yep, unit tests. Oh, and CppCheck alerts us about these, too; I will have to find a way to calm it down about the places where using strncpy is exactly what we want and that yes, we do intend the behavior that CppCheck identifies as risky ;)

**Comment by skliper** [comment]
_Trac comment by jphickey on 2016-02-10 13:36:02:_

Rebased commit at [changeset:be9b52a]

**Comment by skliper** [comment]
_Trac comment by glimes on 2016-02-16 13:40:11:_

Note from CCB review:
    void CFE_ES_HousekeepingCmd(CFE_SB_MsgPtr_t Msg)
replacing memcpy with a loop: two memcpy were present, the replacement
only does the copy for the first one. Update of PerfFilterMask is missing.

**Comment by skliper** [comment]
_Trac comment by jphickey on 2016-02-25 11:00:07:_

Commit [changeset:81ffffd] addresses the above concern.

I have re-reviewed the rest of this changeset and everything else looks OK.

(Note this is not an "amend" commit but rather a follow-on, so the diff above will only show the fix for this, not the rest of the change set).

**Comment by skliper** [comment]
_Trac comment by glimes on 2016-03-09 15:03:27:_

Authorized by CCB, will be merged via ic-2016-03-08
There were some conflicts with a MISRA-motivated change
that indented a block; resolved using MELD and appropriate
flags to ignore whitespace.

**Comment by skliper** [comment]
_Trac comment by jhageman on 2019-03-05 14:57:55:_

Milestone renamed
